### DISCUSSION TITLE
define a single bond connection between a querry atom and any atom (including implicit H) does not work

### DISCUSSION THREAD

**Discussion by cottonkiet** [rdkit/rdkit#5220] [Q&A]
Hi

I am using the following reaction:
`scheme = '[#7;d2:1]-,=[#6;X3:2](-[*,#1:5])-,=[#6;R0:3]-[#7+:4](-[#8-])=O>>[*,#1:5]-[#6:2]=[#7:1]'
`
with one reactant:
`react1 = Chem.MolFromSmiles('CNC=C[N+]([O-])=O')
`

however, the #1 at atom 5 does not seem to include implicit H here, so this reaction does not happen. Is there any way to make the reaction work? 
The reason that I write scheme this way is because I want react 1, 2 and 3 to pass and react 4 to fail. (see following code)

To reproduce
```
from rdkit import Chem
from rdkit.Chem import AllChem
react = Chem.MolFromSmiles('CNC=C[N+]([O-])=O') # react1 should pass but fail
# react = Chem.MolFromSmiles('CNC(C)=C[N+]([O-])=O') #react2 should pass and pass
# react = Chem.MolFromSmiles('CN=C(C)C[N+]([O-])=O') #react3 should pass and pass
# react = Chem.MolFromSmiles('CNC(=C)C[N+]([O-])=O') #react4 should fail and fail

display(react)
scheme = '[#7;d2:1]-,=[#6;X3:2](-[*,#1:5])-,=[#6;R0:3]-[#7+:4](-[#8-])=O>>[*,#1:5]-[#6:2]=[#7:1]'
rxn = AllChem.ReactionFromSmarts(scheme)
display(rxn)
prod = rxn.RunReactants([react])
len(prod)
display(prod[0][0])
```

**Comment by bp-kelley** [comment]
You are running into implicit hydrogen issues.  Hydrogens, by default, don't exist in the search graph so "(-[*,#1:5])" doesn't see them.

The easiest solution is to add hs when you search.

```
from rdkit import Chem
from rdkit.Chem import AllChem
reacts = [Chem.MolFromSmiles('CNC=C[N+]([O-])=O') # react1 should pass but fail
    ,Chem.MolFromSmiles('CNC(C)=C[N+]([O-])=O') #react2 should pass and pass
    ,Chem.MolFromSmiles('CN=C(C)C[N+]([O-])=O') #react3 should pass and pass
    ,Chem.MolFromSmiles('CNC(=C)C[N+]([O-])=O') #react4 should fail and fail
         ]

# add Hs
reacts = [Chem.AddHs(react) for react in reacts]
scheme = '[#7;d2:1]-,=[#6;X3:2](-[*,#1:5])-,=[#6;R0:3]-[#7+:4](-[#8-])=O>>[*,#1:5]-[#6:2]=[#7:1]' 
rxn = AllChem.ReactionFromSmarts(scheme)
for i,react in enumerate(reacts):
    prod = rxn.RunReactants([react])
    if prod:
        print(i, "matches")
```

outputs:

```
0 matches
1 matches
2 matches
```

But even with these, I'm not sure I understand your reaction.

A harder solution is to use recursive smarts, but without understanding the reaction you are attempting I won't hazard a guess.

**Reply by cottonkiet** [reply]
Thanks a lot! This works! 
If there is a solution with recursive smarts, it will be better for my situation. 
The whole idea of my complicated reaction smarts (X3 for atom 2; single/double bond for bond between atoms 2 and 3 and bond between 1 and 2, and single bond between atom 2 and 5) is to allow two situations to pass: [N;d2:1]-[C:2]=[C;R0:3], and [N;d2:1]=[C:2]-[C;R0:3]. I don't really care about atom 5, but need the bond between atom2 and 5 to be single to do the trick. I fail to find a proper recursive smarts to do the trick though. Is there something I have not thought of? Thanks so much!
